File: hena/tyama_hena24_enum.rb
#!/usr/bin/env ruby
#http://qiita.com/Nabetani/items/1c83005a854d2c6cbb69
#http://nabetani.sakura.ne.jp/hena/ord24eliseq/

=begin
if RUBY_VERSION<'1.9'
	module Math
		def self.cbrt(n)
			n**(1-2.0/3)
		end
	end
end
=end

class Integer
	def isqrt
		return 0 if self<=0
		return 1 if self<4 # 1
		x,y=0,self
		while x!=y&&x+1!=y
			x,y=y,(self/y+y)/2
		end
		x
	end
	def icbrt
		return -self.icbrt if self<0
		return 0 if self==0
		return 1 if self<8 # 1,7
		x,y=0,self
		while x!=y&&x+1!=y
			x,y=y,(self/y/y+y*2)/3
		end
		x
	end
end

=begin
def generate
	return to_enum(:generate) if !block_given?
	i=1
	loop{
		yield i
		i+=1
	}
end
=end

def drop_prev(check,prev)
	return to_enum(:drop_prev,check,prev) if !block_given?
	a=prev.next
	b=prev.next
	loop{
		yield a if !check[b]
		a,b=b,prev.next
	}
end
def drop_next(check,prev)
	return to_enum(:drop_next,check,prev) if !block_given?
	a=prev.next
	b=prev.next
	yield a
	loop{
		yield b if !check[a]
		a,b=b,prev.next
	}
end
def drop_n(check,n,prev)
	return to_enum(:drop_n,check,n,prev) if !block_given?
	i=0
	loop{
		i+=1
		a=prev.next
		yield a if !check[i,n]
	}
end

is_sq=lambda{|n|n.isqrt**2==n}
is_cb=lambda{|n|n.icbrt**3==n}
is_multiple=lambda{|i,n|i%n==0}
is_le=lambda{|i,n|i<=n}

if RUBY_VERSION<'1.9'
	f={
		'S'=>lambda{|enum|drop_next(is_sq,enum)},
		's'=>lambda{|enum|drop_prev(is_sq,enum)},
		'C'=>lambda{|enum|drop_next(is_cb,enum)},
		'c'=>lambda{|enum|drop_prev(is_cb,enum)},
		'h'=>lambda{|enum|drop_n(is_le,100,enum)},
	}
	(2..9).each{|e|f[e.to_s]=lambda{|enum|drop_n(is_multiple,e,enum)}}
else
	f={
		'S'=>Kernel.method(:drop_next).to_proc.curry[is_sq],
		's'=>Kernel.method(:drop_prev).to_proc.curry[is_sq],
		'C'=>Kernel.method(:drop_next).to_proc.curry[is_cb],
		'c'=>Kernel.method(:drop_prev).to_proc.curry[is_cb],
		'h'=>Kernel.method(:drop_n).to_proc.curry[is_le,100],
	}
	(2..9).each{|e|f[e.to_s]=Kernel.method(:drop_n).to_proc.curry[is_multiple,e]}
end

if $0==__FILE__
	STDOUT.sync=true
	while gets
		#cS => f['S'].call(f['c'].call(1.upto(1/0.0)))
		puts $_.chomp.chars.reduce(1.upto(1/0.0)){|s,e|f[e][s]}.take(10)*','

		#enum=generate
		#$_.chomp.chars.each{|e|enum=f[e][enum]}
		#puts enum.take(10)*','
	end
end

File: hena/tyama_hena24_enum.pl
#!/usr/bin/env perl
#http://qiita.com/Nabetani/items/1c83005a854d2c6cbb69
#http://nabetani.sakura.ne.jp/hena/ord24eliseq/

use strict vars;
use feature qw(say);
use IO::Handle;
use Generator::Object;
use List::Util qw(reduce);

#sub cbrt{
#	my($n)=@_;
#	return $n**(1-2.0/3);
#}

sub isqrt{
	my($n)=@_;
	if($n<=0){return 0;}
	if($n<4){return 1;}
	my $x=0;my $y=$n;
	for(;$x!=$y&&$x+1!=$y;){$x=$y;$y=int((int($n/$y)+$y)/2);}
	return $x;
}
sub icbrt{
	my($n)=@_;
	if($n<0){return icbrt(-$n);}
	if($n==0){return 0;}
	if($n<8){return 1;}
	my $x=0;my $y=$n;
	for(;$x!=$y&&$x+1!=$y;){$x=$y;$y=int( (int(int($n/$y)/$y)+$y*2)/3 );}
	return $x;
}

sub generate{
	generator{
		my $i=1;
		for(;;){
			$_->yield($i);
			$i++;
		}
	}
}

sub drop_prev{
	my($check,$prev)=@_;
	generator{
		my $a=$prev->next;
		my $b=$prev->next;
		for(;;){
			$_->yield($a) if !$check->($b);
			$a=$b;
			$b=$prev->next;
		}
	}
}
sub drop_next{
	my($check,$prev)=@_;
	generator{
		my $a=$prev->next;
		my $b=$prev->next;
		$_->yield($a);
		for(;;){
			$_->yield($b) if !$check->($a);
			$a=$b;
			$b=$prev->next;
		}
	}
}
sub drop_n{
	my($check,$n,$prev)=@_;
	generator{
		my $i=0;
		for(;;){
			$i+=1;
			$a=$prev->next;
			$_->yield($a) if !$check->($i,$n);
		}
	}
}
sub is_sq{
	my($n)=@_;
	#my $x=int(sqrt($n));
	my $x=isqrt($n);
	return $x*$x==$n;
}
sub is_cb{
	my($n)=@_;
	#my $x=int(cbrt($n));
	my $x=icbrt($n);
	return $x*$x*$x==$n;
}
sub is_multiple{
	my($i,$n)=@_;
	return $i%$n==0;
}
sub is_le{
	my($i,$n)=@_;
	return $i<=$n;
}

my %f=(
	'S'=>sub{my($enum)=@_;return drop_next('is_sq',$enum);},
	's'=>sub{my($enum)=@_;return drop_prev('is_sq',$enum);},
	'C'=>sub{my($enum)=@_;return drop_next('is_cb',$enum);},
	'c'=>sub{my($enum)=@_;return drop_prev('is_cb',$enum);},
	'h'=>sub{my($enum)=@_;return drop_n('is_le',100,$enum);},
);
for(my $i=2;$i<=9;$i++){$f{"$i"}=sub{my($i)=@_;sub{my($enum)=@_;return drop_n('is_multiple',$i,$enum);}}->($i);}

STDOUT->autoflush;
for(;my $line=<>;){
	chomp($line);
	#cS => $f['S']->($f['c']->(generate()))
	my $z=reduce {$f{$b}->($a)} generate(),split('',$line);
	my $cnt=10;
	my @a;
	for(my $i=0;$i<$cnt;$i++){
		push(@a,$z->next);
	}
	say join(',',@a);
}
